### PR TITLE
Custom keadm join cmd for containerized integration on k8s platform

### PR DESCRIPTION
**Pull request by zhu733756**
Signed-off-by: zhu733756 <zhu733756@kubesphere.io>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This feature adds some extra parameters for keadm,  for better-containerized integration on k8s platform.  Therefore, the generated edgecore.yaml configuration can be controlled by these parameters. 

**Which issue(s) this PR fixes**:

Part of #https://github.com/kubeedge/kubeedge/issues/3300

**Special notes for your reviewer**:
outputs for keadm join:
```
Flags:
      --certPath string                  The certPath used by edgecore, the default value is /etc/kubeedge/certs (default "/etc/kubeedge/certs")
  -s, --certport string                  The port where to apply for the edge certificate
      --cgroupdriver string              CGroupDriver that uses to manipulate cgroups on the host (cgroupfs or systemd), the default value is cgroupfs
  -e, --cloudcore-ipport string          IP:Port address of KubeEdge CloudCore
  -z, --download-region string           To distinguish download links for KubeEdge EdgeCore and Services, could be 'zh' or 'en'
  -a, --edgenode-ip string               KubeEdge Node internal IP reported to K8s cluster, If flag not used then the command will use the IP address read from node network
  -i, --edgenode-name string             KubeEdge Node unique identification string, If flag not used then the command will generate a unique id on its own
  -h, --help                             help for join
  -d, --kubeedge-download-url string     The custom download url for KubeEdge tar file and checksum file. i.e: https://github.com/kubeedge/kubeedge/releases/download
      --kubeedge-version string          Use this key to download and use the required KubeEdge version
  -l, --labels strings                   Use this key to set the customized labels for node. you can input customized labels like key1=value1,key2=value2
  -q, --quicport string                  The port where to apply for the edge quic
  -p, --remote-runtime-endpoint string   KubeEdge Edge Node RemoteRuntimeEndpoint string, If flag not set, it will use unix:///var/run/dockershim.sock
  -r, --runtimetype string               Container runtime type
  -c, --service-file-url-format string   The custom download url format for service file. i.e: https://raw.githubusercontent.com/kubeedge/kubeedge/release-%s/build/tools/%s, the first %s means the version without 'v' like 1.8.2, the second %s means the service name like edgecore.service
      --tarballpath string               Use this key to set the temp directory path for KubeEdge tarball, if not exist, download it
  -t, --token string                     Used for edge to apply for the certificate
  -n, --tunnelport string                The port where to apply for the edge stream connected tunnel
  -w, --with-edge-taint                  To determine whether to add NoSchedule taint to the edge node
  -m, --with-meta-server                 To determine whether to enable metaServer component on edge node
  -k, --with-service-bus                 To determine whether to enable serviceBus component on edge node
```

Example:
```
./keadm join --kubeedge-version=1.8.2  --cloudcore-ipport=192.168.88.6:30000 --quicport 30001 --certport 30002 --tunnelport 30004 --edgenode-name ke1 --edgenode-ip 1.1.1.1  --download-region zh --kubeedge-download-url https://kubeedge.pek3b.qingstor.com/releases/download  --service-file-url-format https://kubeedge.pek3b.qingstor.com/releases/service/%s/%s  --token 8e74ba5c6e3117ba03a072add6d381868d02176681b4fc9bea9943d3697c2781.eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2MzczNjE1OTl9.9OTImeNW8Hq4DB3YASRWflyn_UBgRTYGn5c7rJNsGcA --with-edge-taint --with-meta-server
```

And support reading extra flags and values from the env file:
```
quicport=10001
certport=10002
tunnelport=10004
cloudcore-ipport=192.168.88.6:10000
download-region=zh
kubeedge-download-url=https://kubeedge.pek3b.qingstor.com/releases/download
service-file-url-format=https://kubeedge.pek3b.qingstor.com/releases/service/%s/%s
token=e181e46e30d7bea8575c093670ccd27395823cfb88ac2c2a895d8543ec3d6445.eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2MzgzMDYyNjR9.CMB0ZXTYRE-EeyjNChD7igd5NKxd3HI6bh88QMOlvvU
with-edge-taint
with-meta-server
with-service-bus
```
Therefore,  the above cmd can be simplified as follows:
```
./keadmplus join --edgenode-name=ke1  --edgenode-ip=1.1.1.1
Host has mosquit+ already installed and running. Hence skipping the installation steps !!!
kubeedge-v1.8.2-linux-amd64.tar.gz checksum: 
checksum_kubeedge-v1.8.2-linux-amd64.tar.gz.txt content: 
[Run as service] start to download service file for edgecore
[Run as service] success to download service file for edgecore
kubeedge-v1.8.2-linux-amd64/
kubeedge-v1.8.2-linux-amd64/edge/
kubeedge-v1.8.2-linux-amd64/edge/edgecore
kubeedge-v1.8.2-linux-amd64/cloud/
kubeedge-v1.8.2-linux-amd64/cloud/csidriver/
kubeedge-v1.8.2-linux-amd64/cloud/csidriver/csidriver
kubeedge-v1.8.2-linux-amd64/cloud/admission/
kubeedge-v1.8.2-linux-amd64/cloud/admission/admission
kubeedge-v1.8.2-linux-amd64/cloud/cloudcore/
kubeedge-v1.8.2-linux-amd64/cloud/cloudcore/cloudcore
kubeedge-v1.8.2-linux-amd64/version

KubeEdge edgecore is running, For logs visit: journalctl -u edgecore.service -b
```

**Does this PR introduce a user-facing change?**:

```release-note
keadm support extra parameters
```

### CODE REVIEW

File: keadm/cmd/keadm/app/cmd/common/constant.go [hunk 1]
```diff
@@ -65,6 +77,21 @@ const (
...
+	DownloadRegion = "download-region"
+
+	// Tunnel sets the port where to apply for the edge stream
+	TunnelPort = "tunnelport"
+
+	// Custom download url for KubeEdge tar file and checksum file
+	KubeEdgeDownloadURL = "kubeedge-download-url"
+
+	// Custom download url for service file
+	ServiceFileURLFormat = "service-file-url-format"
```
Comment (gy95) [inline keadm/cmd/keadm/app/cmd/common/constant.go:93]: what's this variable used for? I found this is only used as a left value.~.~

File: keadm/cmd/keadm/app/cmd/util/edgecoreinstaller.go [hunk 2]
```diff
@@ -115,16 +137,41 @@ func (ku *KubeEdgeInstTool) createEdgeConfigFiles() error {
...
 	if ku.CertPort != "" {
 		edgeCoreConfig.Modules.EdgeHub.HTTPServer = "https://" + cloudCoreIP + ":" + ku.CertPort
 	} else {
 		edgeCoreConfig.Modules.EdgeHub.HTTPServer = "https://" + cloudCoreIP + ":10002"
 	}
-	edgeCoreConfig.Modules.EdgeStream.TunnelServer = net.JoinHostPort(cloudCoreIP, strconv.Itoa(constants.DefaultTunnelPort))
+	if ku.QuicPort != "" {
+		edgeCoreConfig.Modules.EdgeHub.Quic.Server = cloudCoreIP + ":" + ku.QuicPort
+	} else {
+		edgeCoreConfig.Modules.EdgeHub.Quic.Server = cloudCoreIP + ":10001"
```
Comment (gy95) [inline keadm/cmd/keadm/app/cmd/util/edgecoreinstaller.go:155]: These two IP address combinations can also be replaced with `net.JoinHostPort` function

File: keadm/cmd/keadm/app/cmd/common/constant.go [hunk 3]
```diff
@@ -65,6 +77,24 @@ const (
...
+
+	// Env file contains extra flags for building edgecore setup configuration
+	KeadmJoinFlagsPath = "/etc/kubeedge/keadm-join-flags.env"
```
Comment (gy95) [inline keadm/cmd/keadm/app/cmd/common/constant.go:96]: what's the format of the contents of this document? just like this? 😂 
```
a=b
c=d
```

File: keadm/cmd/keadm/app/cmd/edge/join.go [hunk 4]
```diff
@@ -109,21 +118,150 @@ func addJoinOtherFlags(cmd *cobra.Command, joinOptions *types.JoinOptions) {
...
+	cmd.Flags().StringVarP(&joinOptions.KubeEdgeDownloadURL, types.KubeEdgeDownloadURL, "d", joinOptions.KubeEdgeDownloadURL,
+		"The custom download url for KubeEdge tar file and checksum file. i.e: https://github.com/kubeedge/kubeedge/releases/download")
+
+	cmd.Flags().StringVarP(&joinOptions.ServiceFileURLFormat, types.ServiceFileURLFormat, "c", joinOptions.ServiceFileURLFormat,
+		"The custom download url format for service file. i.e: https://raw.githubusercontent.com/kubeedge/kubeedge/release-%s/build/tools/%s, "+
+			"the first %s means the version number like 1.8, the second %s means the service name like edgecore.service")
 }
 
 // newJoinOptions returns a struct ready for being used for creating cmd join flags.
 func newJoinOptions() *types.JoinOptions {
```
Comment (gy95) [inline keadm/cmd/keadm/app/cmd/edge/join.go:154]: now, we first read key-value pairs from this file, and the command flag will overwrite those pairs? Is this a bit repetitive?

**Comment by Poorunga** [comment]
EdgeMesh needs to change some KubeEdge configurations before using it. Can you consider adding them? (refer to: https://edgemesh.netlify.app/guide/getting-started.html#manual-installation)

1. CloudCore(1 item)
- dynamicController: enable

2. EdgeCore(3 items)
- metaManager.metaServer: enable
- edged.clusterDNS: "169.254.96.16"
- edged.clusterDomain: "cluster.local"

**Comment by zhu733756** [comment]
> 
1 The field `dynamicController: enable` on the cloudside can be controlled by the helm installation of cloudcore.
2 The fields of the edge side will be added.

**Comment by Poorunga** [comment]
@zhu733756 
> 1 The field dynamicController: enable on the cloudside can be controlled by the helm installation process of cloudcore.
2 The fields of the edge side will be added.

OK

File: keadm/cmd/keadm/app/cmd/edge/join.go (hunk 4 above)
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/edge/join.go:154]: I think we can use FTP to send these common parameters, thus the installation of the edge core can be simplified. One day, we can use scripts to join massive edge nodes.

File: keadm/cmd/keadm/app/cmd/common/constant.go (hunk 3 above)
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/common/constant.go:96]: If you have other good ideas, we can make it better :)

File: keadm/cmd/keadm/app/cmd/util/edgecoreinstaller.go (hunk 2 above)
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/util/edgecoreinstaller.go:155]: let me see see.

File: keadm/cmd/keadm/app/cmd/edge/join.go [hunk 5]
```diff
@@ -230,7 +232,23 @@ func newJoinOptions() *types.JoinOptions {
 	if ok {
 		opts.ServiceFileURLFormat = serviceFileURLFormat
 	}
-
+	enableMetaServer, ok := extraFlagsVals[types.EnableMetaServer]
```
Comment (zhu733756) [inline keadm/cmd/keadm/app/cmd/edge/join.go:235]: enableMetaServer should be enableEdgeMesh? /cc @Poorunga

**Comment by zhu733756** [comment]
will refactor keadm in #3475.
/close

**Pull request closed without merging**